File: TypeScript/Problem2.ts
interface CacheNode {
    key: number;
    value: number;
    prev: CacheNode | null;
    next: CacheNode | null;
}

class LRUCache {
    private capacity: number;
    private map: Map<number, CacheNode>;
    private head: CacheNode | null;
    private tail: CacheNode | null;

    constructor(capacity: number) {
        this.capacity = capacity;
        this.map = new Map();
        this.head = null;
        this.tail = null;
    }

    private remove(node: CacheNode): void {
        const prevNode = node.prev;
        const nextNode = node.next;

        if (prevNode) prevNode.next = nextNode;
        if (nextNode) nextNode.prev = prevNode;

        if (this.head === node) this.head = nextNode;
        if (this.tail === node) this.tail = prevNode;
    }

    private insertAtHead(node: CacheNode): void {
        node.next = this.head;
        node.prev = null;

        if (this.head) {
            this.head.prev = node;
        }

        this.head = node;

        if (!this.tail) {
            this.tail = node;
        }
    }

    get(key: number): number {
        if (!this.map.has(key)) {
            return -1;
        }

        const node = this.map.get(key)!;
        this.remove(node);
        this.insertAtHead(node);

        return node.value;
    }

    put(key: number, value: number): void {
        if (this.map.has(key)) {
            const existingNode = this.map.get(key)!;
            this.remove(existingNode);
        }

        const newNode: CacheNode = { key, value, prev: null, next: null };
        this.insertAtHead(newNode);
        this.map.set(key, newNode);

        if (this.map.size > this.capacity) {
            const lru = this.tail!;
            this.remove(lru);
            this.map.delete(lru.key);
        }
    }
}

File: Python/Problem2.py
#LRU Cache
class Node:
    def __init__(self, key, val, prev = None, next = None):
        self.key = key
        self.val = val
        self.prev = prev
        self.next = next

class LRUCache:

    def __init__(self, capacity: int):
        self.map = {}
        self.capacity = capacity
        self.length = 0
        self.startNode = None
        self.endNode = None

    def get(self, key: int) -> int:
        if key in self.map.keys():
            node = self.map[key]
            self.pushToLast(node)
            return self.map[key].val
        return -1

    def put(self, key: int, value: int) -> None:
        if self.length > 0 and key in self.map.keys():
            node = self.map[key]
            node.val = value
            self.pushToLast(node)
            return
        self.insertNode(key, value)

    def pushToLast(self, node: Node):
        if node.next == None:
            return

        if node.prev == None:
            self.startNode = node.next

        if node.prev == None:
            node.next.prev = None
        else:
            node.prev.next = node.next
            node.next.prev = node.prev

        node.prev = self.endNode
        node.next = None
        self.endNode.next = node
        self.endNode = node

    def soloNode(self, node: Node):
        self.startNode = node
        self.endNode = node
    
    def insertNode(self, key: int, value: int):
        node = Node(key, value)
        if self.length < self.capacity:
            if self.length == 0:
                self.soloNode(node)

            else:
                self.endNode.next = node
                node.prev = self.endNode
                self.endNode = node

            self.map[key] = node
            self.length += 1
        else: 
            del self.map[self.startNode.key]
            if self.capacity == 1:
                self.soloNode(node)
            else:
                
                self.startNode = self.startNode.next
                self.startNode.prev = None

                
                self.endNode.next = node
                node.prev = self.endNode
                self.endNode = node
            
            self.map[key] = node
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)